quil-rs/src/instruction/mod.rs

```diff
@@ -410,6 +410,18 @@ pub fn format_qubits(qubits: &[Qubit]) -> String {
         .join(" ")
 }
 
+/// Format qubits as a Quil parameter list, where each variable qubit must be prefixed with a `%`.
+pub fn format_qubit_parameters(qubits: &[Qubit]) -> String {
+    qubits
+        .iter()
+        .map(|q| match q {
+            Qubit::Fixed(index) => format!("{index}"),
+            Qubit::Variable(var) => format!("%{var}"),
+        })
+        .collect::<Vec<String>>()
+        .join(" ")
+}
+
```

Comment:
This is an edited version of @kalzoo's suggestion that I think is more likely to apply cleanly.

```suggestion
fn write_qubit_parameters(f: &mut fmt::Formatter, qubits: &[Qubit]) -> fmt::Result {
    let mut qubits = qubits.iter();
    let fmt_qubit = |f, qubit| {
        match qubit {
            Qubit::Fixed(index) => write!(f, "{index}"),
            Qubit::Variable(var) => write!(f, "%{var}"),
        }
    };
    
    if let Some(first) = qubits.first() {
      fmt_qubit(f, first)?;
    }
    
    for qubit in qubits {
        write!(f, " ")?;
        fmt_qubit(f, qubit)?;
    }
    
    Ok(())
}

```

And because it was bugging me, one more alternative:


```suggestion
fn write_qubit_parameters(f: &mut fmt::Formatter, qubits: &[Qubit]) -> fmt::Result {
    for (i, qubit) in qubits.iter().enumerate() {
        if i > 0 {
            write!(f, " ")?;
        }
        match qubit {
            Qubit::Fixed(index) => write!(f, "{index}"),
            Qubit::Variable(var) => write!(f, "%{var}"),
        }
    }
    
    Ok(())
}

```

Reply:
I applied these, with a tweak to simplify things. Instead of treating the first qubit as a special case, we just apply a space before every qubit. The benefit here is two fold:

1) `write_qubit_parameters` is simpler
2) The display implementations for each instruction that uses `write_qubit_parameters` is simpler, since it doesn't have to check if there is at least one qubit and add a space before the parameter list if so.

Since `write_qubit_parameters` now takes the `Formatter`, I think this is a fair abstraction to make, it will correctly format the qubit parameter list, whether it's empty or not.